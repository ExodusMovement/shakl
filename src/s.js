import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
