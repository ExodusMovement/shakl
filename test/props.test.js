import React from 'react'
import { View } from 'react-native'
import { create as r } from 'react-test-renderer'

import s from '../src/rn'

it('accepts a style prop', () => {
  const Foo = s.View({ margin: 10 })
  const foo = r(<Foo style={{ flex: 1 }} />).toJSON()

  expect(foo.props.style).toEqual({ margin: 10, flex: 1 })

  expect(foo).toMatchSnapshot()
})

const theme = {
  fancyColor: '#193bc3',
}

it("doesn't pass `theme` property by default", () => {
  const Foo = s.View(({ theme }) => ({ color: theme.fancyColor }))
  const foo = r(<Foo theme={theme} />).toJSON()

  expect(foo.props.style).toEqual({
    color: '#193bc3',
  })

  expect(foo).toMatchSnapshot()
})

it('passes `theme` property if `omitProps` specified without it', () => {
  const Foo = s(View, { omitProps: [] })(({ theme }) => ({ color: theme.fancyColor }))
  const foo = r(<Foo theme={theme} />).toJSON()

  expect(foo.props.style).toEqual({
    color: '#193bc3',
  })

  expect(foo).toMatchSnapshot()
})
