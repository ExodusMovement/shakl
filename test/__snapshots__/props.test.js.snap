// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts a style prop 1`] = `
<View
  style={
    {
      "flex": 1,
      "margin": 10,
    }
  }
/>
`;

exports[`doesn't pass \`theme\` property by default 1`] = `
<View
  style={
    {
      "color": "#193bc3",
    }
  }
/>
`;

exports[`passes \`theme\` property if \`omitProps\` specified without it 1`] = `
<View
  style={
    {
      "color": "#193bc3",
    }
  }
  theme={
    {
      "fancyColor": "#193bc3",
    }
  }
/>
`;
